content/technotes/2020-04-30-code-highlighting/index.Rmd

```diff
@@ -0,0 +1,411 @@
+<!--/html_preserve-->
+
+The html of the block seen above is
+
+```html
+<div class="highlight"><pre class="chroma"><code class="language-r" data-lang="r"><span class="nf">install.packages</span><span class="p">(</span><span class="s">&#34;parzer&#34;</span><span class="p">,</span> <span class="n">repos</span> <span class="o">=</span> <span class="s">&#34;https://dev.ropensci.org/&#34;</span><span class="p">)</span>
+</code></pre></div>
+```
+
+and it goes hand in hand with having styling for different ".chroma" classes in our website CSS.
+
+```css
+.chroma .s { color: #a3be8c }
+```
+
+To have this behaviour, in our website config there's 
+
+```toml
+pygmentsUseClasses=true
+```
+
+which confusingly enough uses the name "Pygments", not Chroma, for historical reasons.
+You'd use CSS like we do if none of [Chroma default styles](https://xyproto.github.io/splash/docs/all.html) suited you, or if, like we did, had some special highlighting that you'd like to keep.[^css]
+
+
+How does Chroma know what parts of code is of the string class for instance?
+Once again, regular expressions help, in this case in what is called a lexer. 
+Chroma is inspired by Pygments, and in [Pygments docs](https://pygments.org/docs/quickstart/) it is explained that _"A lexer splits the source into tokens, fragments of the source that have a token type that determines what the text represents semantically (e.g., keyword, string, or comment)."_
+In [R lexer](https://github.com/alecthomas/chroma/blob/master/lexers/r/r.go), ported from Pygments to Chroma by Chroma maintainer [Alec Thomas](https://github.com/alecthomas), for strings we e.g. see
+
+```go
+			{`\'`, LiteralString, Push("string_squote")},
+			{`\"`, LiteralString, Push("string_dquote")},
+// ... code
+		"string_squote": {
+			{`([^\'\\]|\\.)*\'`, LiteralString, Pop(1)},
+		},
+		"string_dquote": {
+			{`([^"\\]|\\.)*"`, LiteralString, Pop(1)},
+		},
+```
+
+Chroma works on Markdown content, so if you use [blogdown to generate pages as html](https://bookdown.org/yihui/blogdown/output-format.html), you can only use client-side highlighting, like [this tidyverse.org page](https://www.tidyverse.org/blog/2017/09/erratum-tidyr-0.7.0/) whose [source is html](https://github.com/tidyverse/tidyverse.org/blob/7ceea9b9bd40207d25e75681041568d600c9d903/content/blog/2017-09-erratum-tidyr/index.html).
+By default nowadays Hugo does server-side syntax highlighting but you could choose to turn it off via [`codeFences = false`](https://gohugo.io/getting-started/configuration-markup#highlight).
+
+#### Accessible highlighting?
+
+When selecting a style for code highlighting on your website, or when creating your own CSS classes, it is important to keep accessibility in mind.
+For instance, comments could be darker than code, but their being too dark might make contrast between code and background too low!
```

Comment:
This section definitely feels a bit out of place. I think our idea of having a section at the end of the post is a good idea. That way you can wrap up with accessibility as something to bear in mind.